python/ray/remote_function.py

```diff
@@ -0,0 +1,158 @@
+
+    Attributes:
+        _function: The original function.
+        _function_id: The ID of the function.
+        _function_name: The module and function name.
+        _num_cpus: The default number of CPUs to use for invocations of this
+            remote function.
+        _num_gpus: The default number of GPUs to use for invocations of this
+            remote function.
+        _resources: The default custom resource requirements for invocations of
+            this remote function.
+        _num_return_vals: The default number of return values for invocations
+            of this remote function.
+        _max_calls: The number of times a worker can execute this function
+            before executing.
+        _function_signature: The function signature.
+    """
+
+    def __init__(self, function, num_cpus, num_gpus, resources,
+                 num_return_vals, max_calls):
+        self._function = function
+        # TODO(rkn): We store the function ID as a string, so that
+        # RemoteFunction objects can be pickled. We should undo this when
+        # we allow ObjectIDs to be pickled.
+        self._function_id = compute_function_id(self._function).id()
+        self._function_name = (
+            self._function.__module__ + '.' + self._function.__name__)
+        self._num_cpus = (DEFAULT_REMOTE_FUNCTION_CPUS
+                          if num_cpus is None else num_cpus)
+        self._num_gpus = num_gpus
+        self._resources = resources
+        self._num_return_vals = (DEFAULT_REMOTE_FUNCTION_NUM_RETURN_VALS if
+                                 num_return_vals is None else num_return_vals)
+        self._max_calls = (DEFAULT_REMOTE_FUNCTION_MAX_CALLS
+                           if max_calls is None else max_calls)
+
+        ray.signature.check_signature_supported(self._function)
+        self._function_signature = ray.signature.extract_signature(
+            self._function)
+
+        # # Export the function.
+        worker = ray.worker.get_global_worker()
+        if worker.mode in [ray.worker.SCRIPT_MODE, ray.worker.SILENT_MODE]:
+            self._export()
+        elif worker.mode is None:
+            worker.cached_remote_functions_and_actors.append(
+                ("remote_function", self))
+
+    def __call__(self, *args, **kwargs):
+        raise Exception("Remote functions cannot be called directly. Instead "
+                        "of running '{}()', try '{}.remote()'.".format(
+                            self._function_name, self._function_name))
+
+    def remote(self, *args, **kwargs):
+        """This runs immediately when a remote function is called."""
+        return self._submit(args=args, kwargs=kwargs)
+
+    def _submit(self,
+                args=None,
+                kwargs=None,
+                num_return_vals=None,
+                num_cpus=None,
+                num_gpus=None,
+                resources=None):
+        """An experimental alternate way to submit remote functions."""
```

Comment:
given that this is marked experimental, is this used?

Reply:
It's used in Pandas on Ray (to change the number of return values), and I think it will be used more in the future (e.g., to specify different resource requirements at runtime).